### PR TITLE
Separate the csv adapter from the query building logic

### PR DESCRIPTION
**Pull request by dlindhol**
This depends on the latis3 dataset-as-trait branch (which has been recently rebased on master) but the key elements are here. 

I've separated the query building logic into the `HapiAdapter`, an abstract class whose subclasses must provide the dataset format string (for the query: csv, json, binary) and an Adapter to use for parsing the data. The `HapiCsvAdapter` extends it to specify "csv" and uses the TextAdapter.

We should be able to refactor the HAPI JSON adapter in a similar way where it then becomes the `parsingAdapter`.

### CODE REVIEW

File: src/test/scala/latis/input/HapiAdapterSpec.scala [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+  
+  val baseUri = new URI("http://lasp.colorado.edu/lisird/hapi/")
+  
+  val dataset = new AdaptedDataset(
+    Metadata("sorce_tsi"),
+    model,
+    adapter,
+    baseUri
+  )
+//  val dataset = Dataset.fromName("sorce_tsi") //get from fdml
```
Comment (dlindhol) [inline src/test/scala/latis/input/HapiAdapterSpec.scala]: Any thoughts on defining the dataset in code vs using fdml?

File: src/test/scala/latis/input/HapiAdapterSpec.scala [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+  
+  
+  "A hapi csv request with time selections" should "return csv records" in {
+    val ds = dataset.withOperation(Selection("time", ">", "2010-01-01"))
```
Comment (dlindhol) [inline src/test/scala/latis/input/HapiAdapterSpec.scala]: Note the new `withOperation` that the new dataset stuff supports. The DSL could add a `select` method on `Dataset`.

File: src/test/scala/latis/input/HapiAdapterSpec.scala [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+                    .withOperation(Selection("time", "<", "2011-01-01"))
+    
+    StreamUtils.unsafeHead(ds.samples) match {
```
Comment (dlindhol) [inline src/test/scala/latis/input/HapiAdapterSpec.scala]: Always looking for convenience methods to make life easier. Maybe we could import a test DSL that adds unsafe things to `Dataset` like `head` that would immediately give us the first `Sample`?

File: src/test/scala/latis/input/HapiAdapterSpec.scala (hunk 1 above)
Comment (sapols) [reply src/test/scala/latis/input/HapiAdapterSpec.scala]: Sorry, I haven't actually used an FDML before. What would it look like defined in FDML?

Comment (dlindhol) [reply src/test/scala/latis/input/HapiAdapterSpec.scala]: Very similar to tsml: https://github.com/latis-data/latis3/blob/master/core/src/test/resources/datasets/data.fdml
I'm just not sure if we want a bunch of test fdml files floating around or to define test datasets in code closer to the test.

Comment (sapols) [reply src/test/scala/latis/input/HapiAdapterSpec.scala]: If it's just for the tests then I lean toward defining the dataset in code. That way when I'm going to look at a test, I don't also have to look in an FDML to see the dataset.

Comment (sapols) [reply src/test/scala/latis/input/HapiAdapterSpec.scala]: If it's just for the tests then I lean toward defining the dataset in code. That way when I'm going to look at a test, I don't also have to look in an FDML to see the dataset.

File: src/main/scala/latis/input/HapiAdapter.scala [hunk 4]
```diff
@@ -4,40 +4,103 @@ import latis.model._
...
+          case '=' => ??? //TODO: support single time selection: add epsilon, take first
+          case _ => 
+            val msg = s"Unsupported select operator: $op"
+            throw new UnsupportedOperationException(msg)
+        }
+      //case Projection(vids @ _*) => vids.mkString("parameters=", ",", "")
+      case _ => ??? //TODO: bug, we should only be given Ops that pass canHandleOperation
+    }
+    //TODO: if we don't have both, use defaults from info header
+    //TODO: add epsilon if times are the same? (CDAWeb seems to require it)
```
Comment (sapols) [inline src/main/scala/latis/input/HapiAdapter.scala]: I do think adding an epsilon would be helpful.

File: src/main/scala/latis/input/HapiAdapter.scala [hunk 5]
```diff
@@ -4,40 +4,103 @@ import latis.model._
...
+        op.head match {
+          case '>' => "time.min=" + time
+          case '<' => "time.max=" + time
+          case '=' => ??? //TODO: support single time selection: add epsilon, take first
+          case _ => 
+            val msg = s"Unsupported select operator: $op"
+            throw new UnsupportedOperationException(msg)
+        }
+      //case Projection(vids @ _*) => vids.mkString("parameters=", ",", "")
+      case _ => ??? //TODO: bug, we should only be given Ops that pass canHandleOperation
```
Comment (lindholc) [inline src/main/scala/latis/input/HapiAdapter.scala]: Should we throw an exception here?

File: src/main/scala/latis/input/HapiAdapter.scala [hunk 6]
```diff
@@ -4,40 +4,103 @@ import latis.model._
...
+  
+  def buildQuery(ops: Seq[Operation]): String = {
+    // Define the format of time strings used by the HAPI API.
+    //val timeFormat = TimeFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'")
+    
+    // Placeholder for min time
+    var startTime: Option[String] = None
+    
+    // Placeholder for max time
+    var endTime:   Option[String] = None
```
Comment (lindholc) [inline src/main/scala/latis/input/HapiAdapter.scala]: These variables don't appear to be used anywhere.

File: src/main/scala/latis/input/HapiAdapter.scala [hunk 7]
```diff
@@ -4,40 +4,103 @@ import latis.model._
 import latis.util._
 import java.net.URI
 import latis.data.SampledFunction
-import latis.data.HapiFunction
+import latis.ops.Operation
+import latis.ops.Selection
+import latis.ops.Projection
```
Comment (lindholc) [inline src/main/scala/latis/input/HapiAdapter.scala]: This import isn't used.

File: src/test/scala/latis/input/HapiAdapterSpec.scala [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+    model,
+    new HapiAdapter.Config(
+      "class" -> "latis.input.HapiCsvAdapter",
+      "id"    -> "sorce_tsi_24hr_l3"
+    )
+  )
+  
+  val baseUri = new URI("http://lasp.colorado.edu/lisird/hapi/")
+  
+  val dataset = new AdaptedDataset(
```
Comment (lindholc) [inline src/test/scala/latis/input/HapiAdapterSpec.scala]: To reduce namespace clutter, you could make the dataset like this:

```scala
val dataset = {
  val model = ???
  val adapter = ???
  val baseUri = ???

  new AdaptedDataset(...)
}
```

File: src/main/scala/latis/input/HapiAdapter.scala (hunk 6 above)
Comment (dlindhol) [reply src/main/scala/latis/input/HapiAdapter.scala]: These will be used to support the TODOs, to which I should probably do.

File: src/main/scala/latis/input/HapiAdapter.scala (hunk 7 above)
Comment (dlindhol) [reply src/main/scala/latis/input/HapiAdapter.scala]: Hmm, eclipse is no longer warning me of unused imports. Now if we can figure out how to get IDEs and/or scalafmt to do this for us....

**Comment by dlindhol** [comment]
I handled some of the TODOs and cleaned up a few things. The only bit I'd like to get in here for this PR is the use of time coverage from the HAPI "info" for default time selections. I might borrow something from Chris' HapiReader branch. Projection push down might be nice but we can live without that.

I also changed the test to use a dataset that actually works.

**Pull request merged**